### PR TITLE
Add a C++ Filter Model class

### PR DESCRIPTION
**Pull request by tronical**
This tries to match the FilterModel in the Rust API.

### CODE REVIEW

**Comment by tronical** [comment]
Let's start with this. I'm curious what you think of the API.

**Review by ogoffart — COMMENTED** [review]
>  I'm curious what you think of the API.

Just the same as the Rust API as far as i can tell, so looks good to me.

**Review by ogoffart — APPROVED** [review]

File: api/cpp/include/slint.h [hunk 1]
```diff
@@ -828,6 +828,151 @@ class VectorModel : public Model<ModelData>
     }
 };
 
+template<typename ModelData>
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/include/slint.h:831]: Yes, agreed. (separate PR)

File: api/cpp/include/slint.h [hunk 2]
```diff
@@ -828,6 +828,151 @@ class VectorModel : public Model<ModelData>
...
+    }
+
+    std::shared_ptr<slint::Model<ModelData>> source_model;
+    std::function<bool(const ModelData &)> filter_fn;
+    std::vector<int> accepted_rows;
+    slint::FilterModel<ModelData> &target_model;
+};
+}
+
+/// The FilterModel acts as a proxy model for a given source model by applying a filter
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/include/slint.h:932]: Ah yes, I agree. Will change.

File: api/cpp/include/slint.h [hunk 3]
```diff
@@ -828,6 +828,151 @@ class VectorModel : public Model<ModelData>
     }
 };
 
+template<typename ModelData>
+class FilterModel;
+
+namespace private_api {
+template<typename ModelData>
+struct FilterModelInner : private_api::AbstractRepeaterView
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/include/slint.h:837]: I'll make a commit to rename it to `ModelChangeListener`, like in Rust.

**Pull request merged**